### PR TITLE
feat(enhancement): Flagship model auto condition

### PR DESCRIPTION
**Pull request by warp-core**
**Feature:** This PR implements the feature request detailed and discussed in issue #{{insert number}}
@Zitchas asked for it on Discord.

## Feature Details
Adds an auto condition: `"flagship: <flagship model>"`.

## UI Screenshots
N/A

## Usage Examples
```
has "flagship: Peregrine"
```

## Testing Done
It built and ran.
The expected condition appears in my savefile.

## Performance Impact
N/A

### CODE REVIEW

**Comment by Zitchas** [comment]
This feature will be used immediately in #6761 once implemented.

Thank you!

Question: If someone captures a variant of a ship; will it list the variant name or the base ship name?

For example, if someone captured a `"Albatross (Sniper)"`, will it list that, or will it list `Albatross`?

**Comment by Zitchas** [comment]
Hmmm... Possibly relevant: https://github.com/endless-sky/endless-sky/pull/5244

**Comment by warp-core** [comment]
Base name, not variant name.

Possibly may not go all the way to the base if you use a variant as the base of a variant, but I think even then it'll be fine. I can check tomorrow to be sure.

**Comment by quyykk** [comment]
> For example, if someone captured a "Albatross (Sniper)", will it list that, or will it list Albatross?

The latter.

File: source/PlayerInfo.cpp
Comment (petervdmeer) [inline source/PlayerInfo.cpp:2602]: `flagship` could be a nullptr (if the flagship was destroyed in-flight). That is why the code on lines 2604 and 2615 (original numbering) checks if the flagship pointer is valid before de-referencing it.

**Review by petervdmeer — CHANGES_REQUESTED** [review]

File: source/PlayerInfo.cpp
Comment (warp-core) [reply source/PlayerInfo.cpp]: I think 6e036e2 fixes this, though @Amazinite isn't interested in merging this because of #5244 so I'm not sure it matters.

**Review by petervdmeer — APPROVED** [review]
This PR already looks good to me as-is, but it would be helpful if the new condition is also added to some (integration) tests. The [sales tests](https://github.com/warp-core/endless-sky/blob/Flagship-model-auto-condition/data/tests/tests_ship_sale.txt) for example currently checks the total numbers of ships, but it should not be too hard to also check there if the correct/expected flagship model was found.
Checking the flaghip-model not just at the end, but also in some test directly after loading could also be valuable to check that the model also initially gets set correctly. (But I'm not sure if #5244 or a similar change is needed to get that working properly.)
If such tests are added, then merging this PR ahead of #5244 would make quite a bit of sense, since the new tests would also provide test-coverage for #5244.

File: source/PlayerInfo.cpp
Comment (Amazinite) [reply source/PlayerInfo.cpp]: My stance has essentially been that once #5244 is in, then we can start going crazy with new auto-conditions if we want (e.g. visited planets/systems, more fleet information like flagship info or fleet size, installed outfits (#5412)).

**Comment by warp-core** [comment]
> Checking the flaghip-model not just at the end, but also in some test directly after loading could also be valuable to check that the model also initially gets set correctly. (But I'm not sure if #5244 or a similar change is needed to get that working properly.) If such tests are added, then merging this PR ahead of #5244 would make quite a bit of sense, since the new tests would also provide test-coverage for #5244.

It doesn't seem like UpdateAutoConditions runs on Load (maybe it should?) so testing before taking off would just be checking that the value was loaded from the save file correctly, which seems a little unnecessary since, if that isn't loaded correctly, there's probably a bigger problem (loading it should be no different to loading any other condition), and if it doesn't exist in the save file, it'll fail. I can still write those tests for when #5244 is done, though.

**Comment by petervdmeer** [comment]
> It doesn't seem like UpdateAutoConditions runs on Load (maybe it should?) so testing before taking off would just be checking that the value was loaded from the save file correctly, which seems a little unnecessary <...>

I agree, let's keep this out of scope of this PR.

**Comment by petervdmeer** [comment]
As Amazinite indicated; we will add more flagship related info in the conditions after #5244 gets merged. Would it make sense to use `flagship model:` as beginning of the keyword instead of `flagship:` (so that we can also add keywords like `flagship category:` and others).

**Comment by warp-core** [comment]
Done

**Comment by warp-core** [comment]
Oh, right, that broke the tests.

While I'm here, is it right to just tack these tests onto the end of the ship selling ones?

**Comment by petervdmeer** [comment]
> While I'm here, is it right to just tack these tests onto the end of the ship selling ones?

Yes, that was fine, but I still have a small request for "negatives testing", so let's not only check `"flagship model: Finch" == 1` in the first test, but also check things like `"flagship model: Bactrian" == 0` in that first test (to also test/verify the clear code).

**Comment by warp-core** [comment]
I added them but don't think these actually test clearing (unless UpdateAutoConditions runs when you sell a ship?) because these tests don't take off before the flagship changes, so there was nothing to clear.

**Comment by warp-core** [comment]
So, the new test I added fails. Investigating it, I think it's because the auto condition updates that occur on depart (because IncrementDay occurs) aren't written to the savefile at that point. Flagship required crew also doesn't update in the savefile.

**Comment by warp-core** [comment]
Actually, I think it might just be the same issue as in Sell 0 Ships.

**Comment by warp-core** [comment]
Ok, I don't know what's wrong.

File: data/tests/tests_common.txt
Comment (petervdmeer) [inline data/tests/tests_common.txt:594]: This function sends the key-input for landing, but does not actually have the test-code wait until the ship has landed. (The watchdog is a "safety timeout" to avoid hanging forever in a hanging test, but it is not a wait-loop.) Any inputs given after this land-function will not happen on-planet, but will likely be given while the ship still is busy landing.

I usually use the `"flagship planet: <planetname>"` condition together with a branch to [wait until the flagship has landed](https://github.com/warp-core/endless-sky/blob/4d4f4a1a4099cf82a5165d1a87d012e5256b768b/data/tests/tests_framework.txt#L121).

**Comment by petervdmeer** [comment]
The test-output tells me that the last assert from `test "Clear flagship model condition"` fails because the very first Finch apparently is not sold yet at that point ("Flame Meteor" should have been the very first ship that was sold, but somehow it is still the flagship when the last assert happens):

From the test-output:
```
 #     Call-stack:
#     - "Clear flagship model condition", step: 11 (assert)
#     
#     flagship name: Flame Meteor, system: Rutilicus, planet: <not set>, hull: 1, shields: 1, energy: 1, fuel: 1, heat: 0
#     escort name: Sidewinder Tracker Carrier, system: Rutilicus, planet: <not set>, hull: 1, shields: 1, energy: 1, fuel: 1, heat: 0.32
#     escort name: Sidewinder Carrier, system: Rutilicus, planet: <not set>, hull: 1, shields: 1, energy: 1, fuel: 1, heat: 0.41
#     escort name: Escort 1, system: Sol, planet: <not set>, hull: 1, shields: 1, energy: 1, fuel: 1, heat: 0.22
#     escort name: Escort 2, system: Sol, planet: <not set>, hull: 1, shields: 1, energy: 1, fuel: 0.7, heat: 0.21
```
This is likely not an issue in your code.

**Comment by petervdmeer** [comment]
My expectation is that the test fails because this warning pops up (when launching with one of the Finches as flagship):
warning_popup

The easiest fix for this would be to sell both finches before the first depart. (And only check the model on the Bactrian and on the Aerie.)

**Comment by petervdmeer** [comment]
Another possible fix is to give both Finches a hyperdrive inside the test-savegame. :)

**Comment by warp-core** [comment]
Curious that the sell ship test doesn't have this issue. Does IncrementDate run before that popup happens?

**Comment by petervdmeer** [comment]
> Curious that the sell ship test doesn't have this issue. Does IncrementDate run before that popup happens?

I currently don't have time to investigate this into detail, but I expect that we have a couple of bugs interacting here:
- The sell 0 ships test is already disabled due to another (reported) bug.
- The pop-up for "sell 1 ships" does not seem to show up directly after loading, only after landing and then departing (this is likely a new bug; the flagship is probably seen as "Carryable" during the flight-checks directly after loading, but as "Non-Carryable" after departing and landing. Carryable ships don't trigger this pop-up, since they can be carried.)
- The "sell 2 ships" (and more) tests should not be affected, since both finches will be sold before first departure.

The tests can be made a lot more robust if we have an automatically set condition that tells the test-framework if the player is on a planet or in active flight (which I wanted to do through #5244). Because then we can detect inside the "Depart" test-function if it correctly transitions from "being on a planet" to "being in flight".

**Comment by petervdmeer** [comment]
I had a brief look at your test, and noticed that the `"flagship planet: New Boston"` condition still is set in the conditions after "Depart", causing the "Land" function to return immediately instead of returning after landing. This is likely caused by `UpdateAutoConditions` getting called at departure before the planet was cleared from the flagship.

PR #5244 would also solve this, because the automatic condition would be calculated "on-demand" instead of "cached". If we want to solve this ahead of #5422, then we should probably add another call to UpdateAutoConditions() late in the depart sequence.

I am starting to strongly agree with @Amazinite in that we should just merge #5244 first, it will improve our quality of life quite significantly (and will remove the need to file and fix bugs that are already solved by #5244).

**Comment by warp-core** [comment]
Oh, I thought we were giving up on this for now?

**Comment by petervdmeer** [comment]
> Oh, I thought we were giving up on this for now?

Giving up? We all wanted to have this feature merged. The only discussion we had was on the order of merging compared to another PR.

**Comment by warp-core** [comment]
Didn't this still have failing tests?

**Pull request merged**
